distribution/sc-proxy-dex/src/wrapped_farm_token_merge.rs

```diff
@@ -0,0 +1,331 @@
+            .execute_on_dest_context_custom_range(|_, after| (after - 1, after))
+    }
+
+    fn merge_farming_tokens(
+        &self,
+        tokens: &[WrappedFarmToken<Self::BigUint>],
+    ) -> SCResult<GenericTokenAmountPair<Self::BigUint>> {
+        if tokens.len() == 1 {
+            let first_token = tokens[0].clone();
+            let farming_amount = self.rule_of_three(
+                &first_token.token_amount.amount,
+                &first_token.attributes.farm_token_amount,
+                &first_token.attributes.farming_token_amount,
+            );
+
+            return Ok(GenericTokenAmountPair {
+                token_id: first_token.attributes.farming_token_id,
+                token_nonce: first_token.attributes.farming_token_nonce,
+                amount: farming_amount,
+            });
+        }
+
+        let farming_token_id = tokens[0].clone().attributes.farming_token_id;
+        let locked_asset_token_id = self.locked_asset_token_id().get();
+
+        if farming_token_id == locked_asset_token_id {
+            Ok(self.merge_locked_asset_tokens_from_wrapped_farm(tokens))
+        } else {
+            self.merge_wrapped_lp_tokens_from_farm(tokens)
+        }
+    }
+
+    fn merge_wrapped_lp_tokens_from_farm(
+        &self,
+        tokens: &[WrappedFarmToken<Self::BigUint>],
+    ) -> SCResult<GenericTokenAmountPair<Self::BigUint>> {
+        let mut wrapped_lp_tokens = Vec::new();
+
+        for token in tokens.iter() {
+            let wrapped_lp_token_amount = self.rule_of_three(
+                &token.token_amount.amount,
+                &token.attributes.farm_token_amount,
+                &token.attributes.farming_token_amount,
+            );
+            require!(
+                wrapped_lp_token_amount != 0,
+                "Wrapped Lp token amount cannot be zero"
+            );
+
+            let wrapped_lp_token_id = token.attributes.farming_token_id.clone();
+            let wrapped_lp_token_nonce = token.attributes.farming_token_nonce;
+
+            let attributes =
+                self.get_wrapped_lp_token_attributes(&wrapped_lp_token_id, wrapped_lp_token_nonce)?;
+            let wrapped_lp_token = WrappedLpToken {
+                token_amount: GenericTokenAmountPair {
+                    token_id: wrapped_lp_token_id.clone(),
+                    token_nonce: wrapped_lp_token_nonce,
+                    amount: wrapped_lp_token_amount,
+                },
+                attributes,
+            };
+            wrapped_lp_tokens.push(wrapped_lp_token);
+        }
+
+        let merged_locked_token_amount =
+            self.merge_locked_asset_tokens_from_wrapped_lp(&wrapped_lp_tokens);
+        let merged_wrapped_lp_token_amount =
+            self.get_merged_wrapped_lp_tokens_amount(&wrapped_lp_tokens);
+        let lp_token_amount = FftTokenAmountPair {
+            token_id: wrapped_lp_tokens[0].attributes.lp_token_id.clone(),
```

Comment:
Are all the `wrapped_lp_tokens[*].attributes.lp_token_id` values identical? If yes, why?

Reply:
Should have missed the check for that. Added

Reply:
Where is the new check?

Reply:
Just 6 lines above

Reply:
Ah, sorry, I'm not sure why I didn't see it before. I thought I `git pull`ed the last version, but maybe I didn't...

Resolved.

Reply:
The function is finally used by `merge_farming_tokens`, before calling the function, we check all the farm_token_ids are identical. Does it imply that the lp tokens are identical?

Reply:
Not sure what @yzhang90 thinks about this, but, as far as I am concerned, this is resolved: line 294 calls self.require_wrapped_lp_tokens_from_same_pair(&wrapped_lp_tokens)?; , which checks that the lp_tokens are identical.

Of course, depending on the answer to Yi's question, that require might not have been needed.